eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Arbitrary.hs

```diff
@@ -531,22 +532,21 @@ instance (EraPParams era, Arbitrary (PParamsUpdate era)) => Arbitrary (GovProced
   arbitrary =
     GovProcedures <$> arbitrary <*> arbitrary
 
-instance Era era => Arbitrary (ConwayGovPredFailure era) where
-  arbitrary = GovActionsDoNotExist <$> arbitrary
+instance
+  ( Era era
+  , Arbitrary (PParamsHKD StrictMaybe era)
+  ) => Arbitrary (ConwayGovPredFailure era) where
+  arbitrary = genericArbitraryU
```

Comment:
Note: that's the main reason for this PR, as this generators is _rather incomplete_. But digging more into it, I found gap in various other generators (including from previous eras), so I just went ahead and changed them all to use generics.

Reply:
This is not a good solution either. Anything that contains a Map or Set will not generate any elements with Generic. So, unfortunately I cannot accept usage of `generic-random` because we have been bitten by it before. 

This is a really good find though and I really appreciate for pointing it out. So, if you'd be willing to change the implementation in this PR to what I am about to suggest that would be awesome, otherwise someone from the ledger will be able to do that.

The problem that this PR is trying to solve is the lack of help from the compiler when adding new constructors to a sum type. So, instead of falling onto some solution that uses generic magic that has its own issues of similar sort, we can ask the compiler to let us know when there are any changes to the type:

```haskell
  arbitrary = do
    predFailure <-
      oneof [ GovActionsDoNotExist <$> arbitrary
            , MalformedProposal <$> arbitrary]
            ]
    _ <- case predFailure of
      GovActionsDoNotExist {} -> pure ()
      MalformedProposal {} -> pure ()
    pure predFailure
```

Naturally, that has some more boiler plate, but It is faster and safer than relying on generics.

Reply:
> @lehins Anything that contains a Map or Set will not generate any elements with Generic. So, unfortunately I cannot accept usage of generic-random because we have been bitten by it before.

I am not sure to follow you on this and cannot acknowledge of a similar experience. For example, here below is a sample of generated values using those generators; as you can see, maps and lists are not necessarily empty. It's been proving useful to me for catching for a few things at least.

```hs
[ UtxowFailure (ExtraRedeemers [AlonzoMinting 30,AlonzoSpending 27,AlonzoCertifying 27])

, DelegsFailure (DelplFailure (DelegFailure MIRNegativesNotCurrentlyAllowed))

, UtxowFailure (UtxoFailure (BabbageOutputTooSmallUTxO [((Addr Mainnet (KeyHashObj (KeyHash "8tiAsset (fromList [(PolicyID {policyID = ScriptHash "302cbfb50b264b8ccadd4be3d88dde6e043a8e53b9e01a9d60eef96b"},fromList [("736ff0c953cbbb7810b7ec2ac687cb9ec69179b1",1)])])),Datum "@",SJust TimelockScript TimelockConstr AllOf (StrictSeq {fromStrict = fromList [TimelockConstr TimeExpire (SlotNo 8) (blake2b_256: SafeHash "4a97c875e0519d76e17ed47295fbeb8e577745055eb9a14459b9b1c9eec48a0b"),TimelockConstr AnyOf (StrictSeq {fromStrict = fromList [TimelockConstr Signature (KeyHash "65fc709a5e019b8aba76f6977c1c8770e4b36fa76f434efc588747b7") (blake2b_256: SafeHash "682cfccdd768f36de9b4a22d799058d69822ec844cc8a9b1c23349da1cab8df2"),TimelockConstr MOfN 5 (StrictSeq {fromStrict = fromList [TimelockConstr Signature (KeyHash "bd039f956f4b302f3ab6fc7c4bac3350a540f44af81a8492194dd2c2") (blake2b_256: SafeHash "3dfd59a7f8d3ea430af0504780d44d48ff8dd77f72ce4f18817bca5ed6587784"),TimelockConstr Signature (KeyHash "bd039f956f4b302f3ab6fc7c4bac3350a540f44af81a8492194dd2c2") (blake2b_256: SafeHash "3dfd59a7f8d3ea430af0504780d44d48ff8dd77f72ce4f18817bca5ed6587784"),TimelockConstr Signature (KeyHash "a646474b8f5431261506b6c273d307c7569a4eb6c96b42dd4a29520a") (blake2b_256: SafeHash "0a394b35c2c8322bff6a78906aab85f312326ab18df01365b3aeffd30c3c6994"),TimelockConstr Signature (KeyHash "65fc709a5e019b8aba76f6977c1c8770e4b36fa76f434efc588747b7") (blake2b_256: SafeHash "682cfccdd768f36de9b4a22d799058d69822ec844cc8a9b1c23349da1cab8df2"),TimelockConstr Signature (KeyHash "58e1b65718531b42494610c506cef10ff031fa817a8ff75c0ab180e7") (blake2b_256: SafeHash "cc00cdfbcac1c644920677311787b2cb903d12611df25d50df31edb543d9491e")]}) (blake2b_256: SafeHash "d55574c900d2a925c16f095868e7a3ce9860fbf7d0ca5472c97a776b8451d67c"),TimelockConstr AllOf (StrictSeq {fromStrict = fromList [TimelockConstr Signature (KeyHash "4acf2773917c7b547c576a7ff110d2ba5733c1f1ca9cdc659aea3a56") (blake2b_256: SafeHash "f029bce6a1335cc28c14356021b19c2eb905d89c7c8c18d40a14be63842294b0"),TimelockConstr Signature (KeyHash "b16b56f5ec064be6ac3cab6035efae86b366cc3dc4a0d571603d70e5") (blake2b_256: SafeHash "35f87fb67301b2eed8869c3345cb1473bbb7c90ec2f8bc36a41a7746ecf6968c"),TimelockConstr Signature (KeyHash "e0a714319812c3f773ba04ec5d6b3ffcd5aad85006805b047b082541") (blake2b_256: SafeHash "ddafc1e46f2543e8ecba2e24a9daf5bbf7a62426c69c01d9c33a129bd32dbc2a"),TimelockConstr Signature (KeyHash "b5ae663aaea8e500157bdf4baafd6f5ba0ce5759f7cd4101fc132f54") (blake2b_256: SafeHash "53c9391f2bc76c1e1859adc601d0308ff35c7e1a9349711c0d8cd96e93bb6523"),TimelockConstr Signature (KeyHash "e0a714319812c3f773ba04ec5d6b3ffcd5aad85006805b047b082541") (blake2b_256: SafeHash "ddafc1e46f2543e8ecba2e24a9daf5bbf7a62426c69c01d9c33a129bd32dbc2a")]}) (blake2b_256: SafeHash "ef80e733fd8f20e4962cb689377a7658645dc1427b4434446106a68de0f4d9ca"),TimelockConstr AnyOf (StrictSeq {fromStrict = fromList []}) (blake2b_256: SafeHash "d2947a559a27e7f8a00dec9c15211ea769bef3b0aaadb8f439c86063806c9404")]}) (blake2b_256: SafeHash "a937dd08ff572b4b0eec808f3080001e27dd594ba1e5496813d1698290e9e5ea")]}) (blake2b_256: SafeHash "b0d2a6a9faf166ff9798aae99dde55bdb6f76bae3bcd882124b9952eef2ad757")),Coin 131802),((Addr Testnet (KeyHashObj (KeyHash "4430e934e5ef04d7090689254b95b807eae7d271f997de91c0d08e53")) (StakeRefBase (ScriptHashObj (ScriptHash "0ef8527605a46fd7cef492e23ce4be600fc8593bef75632008138d81"))),MaryValue (Coin 3512528835488663582) (MultiAsset (fromList [(PolicyID {policyID = ScriptHash "467f58932b54910584a0e8ea25a225e06a14530b2e96e938c53a3f22"},fromList [("ba04c7525736c1fd24b11504016886f07dbb1ae6a4f74966",2)])])),DatumHash (SafeHash "2c085af6e94870ff4d556ba2e229af1211317c832ccf39a378e516bacd90d49f"),SJust TimelockScript TimelockConstr AnyOf (StrictSeq {fromStrict = fromList [TimelockConstr MOfN 0 (StrictSeq {fromStrict = fromList []}) (blake2b_256: SafeHash "cee80c7d53ddf981bc3f0292cc3cdf6cfc9f2dfcddff2662d1d0f96c76400228"),TimelockConstr AnyOf (StrictSeq {fromStrict = fromList [TimelockConstr MOfN 0 (StrictSeq {fromStrict = fromList []}) (blake2b_256: SafeHash "cee80c7d53ddf981bc3f0292cc3cdf6cfc9f2dfcddff2662d1d0f96c76400228")]}) (blake2b_256: SafeHash "67dc0aedcee10c314c25207b28e00dcb38d2cee94e29ede1e83babe71cccdeb7")]}) (blake2b_256: SafeHash "74f9ec20307944b5cbdaad00b0e4db25e720790d2e4cda71cde79bd5a3479ecd")),Coin 270059)]))

, ConwayWdrlNotDelegatedToDRep (fromList [])

, DelegsFailure (DelegateeNotRegisteredDELEG (KeyHash "bf2826a9709eb6d23b6281fc7997b54d2f827a66159f40fafafc8e71"))

, DelegsFailure (WithdrawalsNotInRewardsDELEGS (fromList [(RewardAcnt {getRwdNetwork = Testnet, getRwdCred = ScriptHashObj (ScriptHash "0813177968a5b9be40c63842b12d402ad92a6400f854b88002f36a2d")},Coin 925334),(RewardAcnt {getRwdNetwork = Testnet, getRwdCred = KeyHashObj (KeyHash "90ebc456a57f39605d335a9dd5fa5f77a906b0592ea9701b00b0f986")},Coin 205962),(RewardAcnt {getRwdNetwork = Testnet, getRwdCred = KeyHashObj (KeyHash "b45ef3d23564ec08fd78a74e5d6906051d9f899e7c5ecb972152f43d")},Coin 216847),(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = ScriptHashObj (ScriptHash "7c2262903248f07d07d218eef38027b5f618f8f1185542fd001a6e9a")},Coin 739823),(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = ScriptHashObj (ScriptHash "ebf177e750382886f282c99801a8b6b593c49171d5d4fc11bed5c62f")},Coin 25937),(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = KeyHashObj (KeyHash "6d3af004f3d97f868660e78fc11b066e089603626eefa5a01fe932d4")},Coin 939608),(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = KeyHashObj (KeyHash "840273085384e4fb612c17fc13bcb19c8504586882ae79d89ef1ec96")},Coin 913645),(RewardAcnt {getRwdNetwork = Mainnet, getRwdCred = KeyHashObj (KeyHash "d037ad4ec85b27e4042584850fee5270006f6bf84ce83a405646b171")},Coin 196430)]))

, ConwayGovFailure (InvalidPolicyHash (SJust (ScriptHash "95ed4716070c98203e21db0c2f18027fb1f31fe4734b8970ddf28926")) (SJust (ScriptHash "0a54ed78e6930c13363f677f61116c8473f096c66662a9ec8323a5f9")))

, DelegsFailure (DelegateeNotRegisteredDELEG (KeyHash "4d3299cb15e20e7cfcfa3124c894ef4d58492b5c4bde100a5f2f8770"))

, ConwayTreasuryValueMismatch (Coin 307287) (Coin 930407)

, ConwayWdrlNotDelegatedToDRep (fromList [])

, DelegsFailure (DelplFailure (DelegFailure (DuplicateGenesisVRFDELEG "a44dfa77d15b61d5839bf68a96e8a72993b25ceb5b3697b1e44d6c9e80515d51")))

]
```

> @lehins So, if you'd be willing to change the implementation in this PR to what I am about to suggest that would be awesome, otherwise someone from the ledger will be able to do that.

I have no intention to change anything, so feel free to close this PR and craft something of your preference. My main intention was to notify about the gaps in the generators, and while at it, provide my current solution. This currently works for me and I am fine with any solution so long as it keeps the generators _complete_.

Reply:
(Note: all arbitrary instances for predicate failures in [Test/Cardano/Ledger/Shelley/Arbitrary.hs](https://github.com/IntersectMBO/cardano-ledger/blob/9e0031e2ff82445692774a1e0721d880281ea2a0/eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Arbitrary.hs) currently rely on `generic-random` & `genericArbitraryU`; which was the source of inspiration for this PR.)

Reply:
You know, I might be misremembering or maybe confusing it with something else. I distinctly remember debugging an issue of QuckChecks Arbitrary instance not generating anything due to generics, but unfortunately I can't remember the details at the moment.

I know that we use `genericArbitraryU` in quite a few places and it worried me, but now I am not so certain that the worry was justified.

In any case, I'll take a closer look at it tomorrow. Thank you for submitting this PR.